src/CloudNative.CloudEvents/CloudEventAttribute.cs

```diff
@@ -0,0 +1,133 @@
+    /// This class represents the abstract concept of an attribute, rather than
+    /// an attribute value.
+    /// </summary>
+    public class CloudEventAttribute
+    {
+        /// <summary>
+        /// The type of the attribute. All values provided must be compatible with this.
+        /// </summary>
+        public CloudEventAttributeType Type { get; }
+
+        /// <summary>
+        /// The name of the attribute. Instances of this class associated with different
+        /// versions of the specification may use different names for the same concept.
+        /// </summary>
+        public string Name { get; }
+
+        /// <summary>
+        /// Indicates whether this attribute is a required attribute.
+        /// Extension attributes are never required.
+        /// </summary>
+        public bool IsRequired { get; }
+
+        /// <summary>
+        /// Indicates whether this attribute is an extension attribute.
+        /// </summary>
+        public bool IsExtension { get; }
+
+        private Action<object> validator;
+
+        // TODO: Have a "mode" of Required/Optional/Extension?
+
+        private CloudEventAttribute(string name, CloudEventAttributeType type, bool required, bool extension, Action<object> validator) =>
+            (Name, Type, IsRequired, IsExtension, this.validator) = (ValidateName(name), type, required, extension, validator);
+
+        internal static CloudEventAttribute CreateRequired(string name, CloudEventAttributeType type, Action<object> validator) =>
+            new CloudEventAttribute(name, type, required: true, extension: false, validator: validator);
+
+        internal static CloudEventAttribute CreateOptional(string name, CloudEventAttributeType type, Action<object> validator) =>
+            new CloudEventAttribute(name, type, required: false, extension: false, validator: validator);
+
+        public static CloudEventAttribute CreateExtension(string name, CloudEventAttributeType type) =>
+            new CloudEventAttribute(name, type, required: false, extension: true, validator: null);
+
+        /// <summary>
+        /// Creates an extension attribute with a custom validator.
+        /// </summary>
+        /// <param name="name"></param>
+        /// <param name="type"></param>
+        /// <param name="validator">Validator to use when parsing or formatting values.
+        /// This is only ever called with a non-null value which can be cast to the attribute type's corresponding
+        /// CLR type. If the validator throws any exception, it is wrapped in an ArgumentException containing the
+        /// attribute details.</param>
+        /// <returns></returns>
+        public static CloudEventAttribute CreateExtension(string name, CloudEventAttributeType type, Action<object> validator) =>
+            new CloudEventAttribute(name, type, required: false, extension: true, validator: validator);
+
+        public override string ToString() => Name;
```

Comment:
Should we add docs?

Reply:
Eventually, yes :)

I tend to find that it's worth leaving "relatively trivial" docs until after we're happy with the design of the code itself... whereas docs that describe non-obvious behavior, e.g. in the CloudEvent indexer, are useful in terms of the mental process. I'm not very consistent though!

Fortunately when we turn on "generate XML docs" we'll get a huge load of warnings for undocumented bits. I definitely, definitely want everything properly documented before we release 2.0. (That means conceptual docs as well as API reference docs, too - even if it's just a directory on GitHub with Markdown files.)